Translation of 'Choosing the state structure' page (#513)
reactjs/fr.react.dev: src/content/learn/choosing-the-state-structure.md

---
title: Choisir la structure de l'état
---

<Intro>

Bien structurer l’état peut créer une différence entre un composant agréable à modifier et débugger, et un qui est une source constante de bugs. Voici des conseils que vous devriez prendre en compte pour structurer l’état.

</Intro>

<YouWillLearn>

* Quand utiliser état unique ou multiple
* Les choses à éviter en organisant l’état
* Résoudre des problèmes connus avec la structure de l’état

</YouWillLearn>

## Les principes pour structurer l’état {/*principles-for-structuring-state*/}

> Quand vous créez un comoposant qui contient des états, vous devrez faire des choix sur le nombre de variables d’état à utiliser et quelle sera la forme de leurs données. Même si il est possible d’écrire des programmes corrects avec une structure d’état peu optimale, il y a quelques principes qui peuvent vous guider pour faire des meilleurs choix :

Reviewer: ```suggestion
Quand vous créez un composant qui contient des états, vous devez faire des choix sur le nombre de variables d’état à utiliser et la forme de leurs données. Même s'il est possible d’écrire des programmes corrects avec une structure d’état sous-optimale, il y a quelques principes qui peuvent vous guider pour faire de meilleurs choix :
```

> 1. **Les états de groupe.** Si vous actualisez tout le temps deux variables d’état ou plus à la fois, essayez de les fusionner en une seule variable d’état.
> 2. **Evitez les contradictions dans l’état.** Quand l’état est stucturé d’une manière où plusieurs parties d’état peuvent être contradictoires et en "désaccord" entre elles, des erreurs peuvent s’induire. Essayez d’éviter cela.
> 3. **Evitez les états redondants.** Si vous pouvez calculer des informations à partir des props du composant ou d’une de ses variables d'état pendant l’affichage, vous ne devriez pas mettre ces informations dans un état du composant.
> 4. **Evitez la duplication d’états.** Quand la même donnée est dupliquée entre plusieurs variables d’état ou dans des objets imbriqués, c’est difficile de les garder synchronisées. Réduisez la duplication quand vous le pouvez.
> 5. **Evitez les états fortement imbriqués** Un état fortement hiérarchisé n’est pas très pratique à actualiser. Quand c’est possible, priorisez une stucture d'état plate.

Reviewer: ```suggestion
1. **Regroupez les états associés.** Si vous mettez tout le temps à jour plusieurs variables d’état à la fois, essayez de les fusionner en une seule variable d’état.
2. **Évitez les contradictions dans l’état.** Quand l’état est structuré de sorte que plusieurs parties d’état puissent être contradictoires, des erreurs peuvent survenir. Essayez d’éviter ça.
3. **Évitez les états redondants.** Si vous pouvez calculer des informations à partir des props du composant ou de ses variables d'état existantes pendant le rendu, vous ne devriez pas mettre ces informations dans un état du composant.
4. **Évitez la duplication d’états.** Quand la même donnée est dupliquée entre plusieurs variables d’état ou dans des objets imbriqués, il est difficile de les garder synchronisées. Réduisez la duplication quand vous le pouvez.
5. **Évitez les états fortement imbriqués.** Un état fortement hiérarchisé n’est pas très pratique à mettre à jour. Quand c’est possible, priorisez une structure d'état plate.
```

Un gros glissement sémantique, et ce "actualiser" qui est plutôt connoté "rafraîchir" (genre depuis un serveur), là où toutes les trads existantes disent "mettre à jour".

Par ailleurs, beaucoup de coquilles ; tu n'as pas fait une relecture avant de finaliser ta PR ?

> Le but derrière ces principes est de *rendre l’état simple à actualiser sans ajouter d’erreurs*. Retirer des données redondantes et dupliquées de l’état aide à s’assurer que toutes ses parties restent synchronisées. C’est similaire à comment un ingénieur de bases de données souhaite ["normaliser" la structure de la base de données](https://docs.microsoft.com/en-us/office/troubleshoot/access/database-normalization-description) pour réduire les risques de bugs. Pour reprendre Albert Einstein, **"Créez votre état le plus simple qu’il puisse être, mais pas plus."**

Reviewer: ```suggestion
Ces principes visent à *rendre l’état simple à actualiser sans créer d’erreurs*. Retirer les données redondantes et dupliquées de l’état aide à s’assurer que toutes ses parties restent synchronisées. C’est un peu comme un ingénieur de bases de données qui souhaite ["normaliser" la structure de la base de données](https://learn.microsoft.com/fr-fr/office/troubleshoot/access/database-normalization-description) pour réduire les risques de bugs. Pour paraphraser Albert Einstein : **« Faites que votre état soit le plus simple possible — mais pas plus simple. »**
```

Pense à traduire les liens quand la VF est dispo.

Maintenant voyons comment ces principes s’appliquent concrètement.

> ## Les états de groupe {/*group-related-state*/}

Reviewer: ```suggestion
## Regrouper les états associés {/*group-related-state*/}
```

> Vous hésitez peut-être quelques fois entre utiliser une variable d’état simple ou multiple.

Reviewer: ```suggestion
Vous hésitez peut-être parfois entre utiliser une ou plusieurs variables d’état.
```

Devriez-vous faire ça ?

```js
const [x, setX] = useState(0);
const [y, setY] = useState(0);
```

Ou ça ?

```js
const [position, setPosition] = useState({ x: 0, y: 0 });
```

> Techniquement, vous pouvez ces deux approches. Mais **si deux variables changent d’état toujours ensemble, une bonne idée serait de les unir en une seule variable d’état.** Vous n’oublirez pas par la suite de toujours les garder synchronisées, comme dans cet exemple où les mouvements du curseur actualisent les 2 coordonnées du point rouge.

Reviewer: ```suggestion
Techniquement, les deux approches sont possibles. Mais **si deux variables d’état changent toujours ensemble, ce serait une bonne idée de les réunir en une seule variable d’état.** Vous n’oublierez ainsi pas ensuite de les garder synchronisées, comme dans cet exemple où les mouvements du curseur mettent à jour les deux coordonnées du point rouge.
```

</Sandpack>

> Un autre cas où vous pouvez regrouper des données dans un objet ou une liste est lorsque vous ne savez pas combien de parties d'état vous aurez besoin. Par exemple, c’est utile quand vous avez un questionnaire dans lequel l’utilisateur peut ajouter des champs personnalisés.

Reviewer: ```suggestion
Une autre situation dans laquelle vous pouvez regrouper des données dans un objet ou une liste, c'est lorsque vous ne savez pas à l'avance de combien d'éléments d'état vous aurez besoin. Par exemple, c’est utile pour un formulaire dans lequel l’utilisateur peut ajouter des champs personnalisés.
```

<Pitfall>

> Si votre variable d’état est un objet, souvenez vous que [vous ne pouvez actualiser qu’un seul de ses champs](/learn/updating-objects-in-state) sans explicitement copier les autres champs. Par exemple, vous ne pouvez pas faire `setPosition({ x: 100 })` dans l’exemple ci-dessus car om n’y aurait plus du tout la variable `y` ! À la place, si vous vouliez définir `x` tout seul, vous feriez soit `setPosition({ ...position, x: 100 })`, soit la séparation en deux variables d’état et `setX(100)`.

Reviewer: ```suggestion
Si votre variable d’état est un objet, souvenez-vous que [vous ne pouvez pas mettre à jour qu’un seul champ](/learn/updating-objects-in-state) sans explicitement copier les autres champs. Par exemple, vous ne pouvez pas faire `setPosition({ x: 100 })` dans l’exemple ci-dessus car il n’y aurait plus du tout la variable `y` ! Au lieu de ça, si vous vouliez définir `x` tout seul, soit vous feriez `setPosition({ ...position, x: 100 })`, soit vous découperiez l'information en deux variables d’état et `setX(100)`.
```

</Sandpack>

> **Maintenant que l’état est "plat" (aussi dit "normalisé"), actualiser des éléments imbriqués devient plus simple**

Reviewer: ```suggestion
**Maintenant que l’état est « plat » (on pourrait aussi dire « normalisé »), mettre à jour des éléments imbriqués devient plus simple.**
```

Désormais, afin d’enlever un lieu vous avez besoin d’actualiser seulement deux niveaux d’état:

- La version actualisée de son lieu *parent* devrait exclure l’ID supprimé de sa liste `childIds`.
- La version actualisée de la "table" d’objet *root* doit inclure la version mise à jour du lieu parent.

> Voici un exemple de commment vous pourriez commencer:

Reviewer: ```suggestion
Voici un exemple de comment vous pourriez procéder :
```

  function handleComplete(parentId, childId) {
    const parent = plan[parentId];
    // Créez une nouvelle version de son lieu parent
    // cela n’inclut pas l’ID de son enfant.
    const nextParent = {
      ...parent,
      childIds: parent.childIds
        .filter(id => id !== childId)
    };

>     // Actulisez l’état de l’objet d’origine...

Reviewer: ```suggestion
    // Mettre à jour l’objet état racine...
```

    setPlan({
      ...plan,

>       // ...pour qu’il ait le parent actualisé

Reviewer: ```suggestion
      // ...pour qu’il ait le parent à jour
```

      <button onClick={() => {
        onComplete(parentId, id);
      }}>

>         Compléter

Reviewer: ```suggestion
        C’est fait
```

</Sandpack>

> Vous pouvez imbriquer des états autant que vous le souhaitez, mais les rendre "plats" peut résoudre beaucoup de problèmes. Cela facilite la mise à jour de l’état, et permet de s’assurer qu’il n’y a pas de duplication dans les différentes parties d’un objet imbriqué.

Reviewer: ```suggestion
Vous pouvez imbriquer des états autant que vous le souhaitez, mais les rendre « plats » peut résoudre de nombreux problèmes. Ça facilite la mise à jour de l’état, et ça permet de s’assurer qu’il n’y a pas de duplication dans les différentes parties d’un objet imbriqué.
```

"Ça", pas "cela" 😉

<DeepDive>

> #### Améliorer l’utilisation de la mémoire {/*improving-memory-usage*/}

Reviewer: ```suggestion
#### Consommer moins de mémoire {/*improving-memory-usage*/}
```

C'est l'objectif plus précis. Franchement la VO devrait dire "Using less memory".

> Idéalement, vous devriez également enlever les éléments supprimés (et leurs enfants !) depuis l’objet "table" pour améliorer l’utilisation de la mémoire. C’est ce que fait cette version. Elle utilise également [Immer](/learn/updating-objects-in-state#write-concise-update-logic-with-immer) pour rendre la logique de mise à jour plus concise.

Reviewer: ```suggestion
Idéalement, vous devriez également enlever les éléments supprimés (et leurs enfants !) depuis l’objet « table » pour consommer moins de mémoire. C’est ce que fait cette version. Elle utilise également [Immer](/learn/updating-objects-in-state#write-concise-update-logic-with-immer) pour rendre la logique de mise à jour plus concise.
```

  function handleComplete(parentId, childId) {
    updatePlan(draft => {

>       // Enlevez des parents l’ID des endroits enfants

Reviewer: ```suggestion
      // Retirer des ID de lieux enfants du parent
```

      const parent = draft[parentId];
      parent.childIds = parent.childIds
        .filter(id => id !== childId);

>       // Oubliez cet endroit et toute sa descendence. 

Reviewer: ```suggestion
      // Oublier cet endroit et tout ce qu'il contient.
```

      <button onClick={() => {
        onComplete(parentId, id);
      }}>

>         Compléter

Reviewer: ```suggestion
        C’est fait
```

</DeepDive>

> Parfois, vous pouvez aussi réduire l’imbrication des états en déplaçant une partie de l’état imbriqué dans les composants enfants. Cela fonctionne bien pour les états éphémères de l’UI qui n’ont pas besoin d’être stockés, comme le fait de savoir si un élément est survolé.

Reviewer: ```suggestion
Parfois, vous pouvez aussi réduire l’imbrication des états en déplaçant une partie de l’état imbriqué dans les composants enfants. C'est bien adapté aux états éphémères de l’UI qui n’ont pas besoin d’être stockés, comme le fait de savoir si un élément est survolé.
```

<Recap>

* Si deux variables d’état sont toujours mises à jour ensemble, envisagez de les fusionner en une seule.
* Choississez soigneusement vos variables d’état pour éviter de créer des "états" impossibles.
* Structurez votre état de manière à réduire les risques d’erreur lors de sa mise à jour.
* Evitez les états dupliqués et redondants afin de ne pas avoir à les synchroniser.
* Ne mettez pas de props *dans* un état à moins que vous ne vouliez spécifiquement empêcher les mises à jour.

> * Pour les évènements d’UI tels que la sélection, conservez l’ID ou l’index dans l’état au lieu de l’objet lui-même.

Reviewer: ```suggestion
* Pour les interactions telles que la sélection d'élément, conservez l’ID ou l’index dans l’état au lieu de l’objet lui-même.
```

<Solution>

> Le problème est que ce composant a un état `color` initialisé avec la valeur initiale de la prop `color`. Mais quand la prop change, cela n’affecte pas la variable d’état ! Donc elles se désynchronisent. Pour régler ce problème, retirez les variables d’état entièrement, et utilisez la prop `color` directement.

Reviewer: ```suggestion
Le problème, c'est que ce composant a un état `color` initialisé avec la valeur initiale de la prop `color`. Mais quand la prop change, ça n’affecte pas la variable d’état ! Donc elles se désynchronisent. Pour régler ce problème, retirez carrément la variable d’état, et utilisez directement la prop `color`.
```

On évite les locutions + "est". On préfère "consiste à", "tient à ce que", "vient de ce que" ou ", c'est que", suivant les cas.

Par ailleurs, comme tu as déjà pu le voir, en français les adverbes sont plus naturellement autour du verbe (avant = épithète, après = attribut) qu'en fin de locution.

</Solution>

> #### Réparer une liste de matériel cassée {/*fix-a-broken-packing-list*/}

Reviewer: ```suggestion
#### Réparer une liste d'affaires cassée {/*fix-a-broken-packing-list*/}
```

Packing list = liste des trucs à fourrer dans ses bagages.

> #### Réparer une liste de matériel cassée {/*fix-a-broken-packing-list*/}

Reviewer: ```suggestion
#### Réparer une liste d'affaires cassée {/*fix-a-broken-packing-list*/}
```

Packing list = liste des trucs à fourrer dans ses bagages.

</Sandpack>

Notez comme les gestionnaires d’évènements sont seulement utiles pour appeler `setItems` après ce changement. Les compteurs d’objets seront maintenant calculés pendant le prochain rendu depuis la liste `items`, et seront donc toujours à jour.

</Solution>

> #### Réparer une sélection qui disparait {/*fix-the-disappearing-selection*/}

Reviewer: ```suggestion
#### Réparer une sélection qui disparaît {/*fix-the-disappearing-selection*/}
```

> Il y a une liste de `letters` dans l’état. Lorsque vous survolez ou cliquez sur un courrier particulier, celui-ci est mis en surbrillance. Le courrier actuellement en surbrillance est stocké dans la variable d’état `highlightedLetter`. Vous pouvez marquer chaque courrier comme "favori" ou "normal", ce qui met à jour la liste `letters` dans l’état.

Reviewer: ```suggestion
Il y a une liste de `letters` dans l’état. Lorsque vous survolez ou cliquez sur un courrier particulier, celui-ci est mis en surbrillance. Le courrier actuellement en surbrillance est stocké dans la variable d’état `highlightedLetter`. Vous pouvez marquer chaque courrier comme "épinglé" ou "normal", ce qui met à jour la liste `letters` dans l’état.
```

Vu les CTA ensuite, on va aller vers "épingler" et "désépingler", je pense. Pas ouf mais bon.

> Ce code fonctionne, mais il y a un bug mineur d’UI. Quand vous appuyez sur "Favori" ou "Retirer", la surbrillance disparait pendant un moment. Cependant, elle réaparait dès que vous déplacez votre curseur ou passez à un autre courrier avec le clavier. Pourquoi cela se produit-il ? Corrigez le problème pour que la surbrillance ne disparaisse pas après le clic sur le bouton.

Reviewer: ```suggestion
Ce code fonctionne, mais il y a un bug mineur d’UI. Quand vous appuyez sur "Épingler" ou "Désépingler", la surbrillance disparaît pendant un moment. Cependant, elle réapparaît dès que vous déplacez votre curseur ou passez à un autre courrier avec le clavier. Que se passe-t-il ? Corrigez le problème pour que la surbrillance ne disparaisse pas après le clic sur le bouton.
```

      <button onClick={() => {
        onToggleStar(letter);
      }}>

>         {letter.isStarred ? 'Retirer' : 'Favori'}

Author: Ici je savais pas trop comment traduire ça proprement

export default function MailClient() {
  const [selectedId, setSelectedId] = useState(null);

>   // A FAIRE : autoriser la sélection multiple

Reviewer: ```suggestion
  // TODO: autoriser la sélection multiple
```

Le marqueur "TODO:" est universel et souvent détecté par les EDI ou leurs extensions (ex. Todo Tree dans VS Code), on va le considérer comme un marqueur technique à ne pas traduire (déjà arbitré dans d'autres pages).

  const selectedCount = 1;

  function handleToggle(toggledId) {
    // A FAIRE : autoriser la sélection multiple
    setSelectedId(toggledId);
  }

  return (
    <>
      <h2>Boîte de réception</h2>
      <ul>
        {letters.map(letter => (
          <Letter
            key={letter.id}
            letter={letter}
            isSelected={

>               // A FAIRE : autoriser la sélection multiple

Reviewer: ```suggestion
              // TODO: autoriser la sélection multiple
```

<Solution>

> À la place d’un simple `selectedId`, enregistrez une *liste* `selectedIds` dans l’état. Par exemple, si vous sélectionnez le premier et le dernier courrier, elle contiendrait `[0, 2]`. Quand rien n’est sélectionné, elle contiendrait une liste vide `[]` :

Reviewer: ```suggestion
Au lieu d’un simple `selectedId`, enregistrez une *liste* `selectedIds` dans l’état. Par exemple, si vous sélectionnez le premier et le dernier courrier, elle contiendrait `[0, 2]`. Quand rien n’est sélectionné, elle contiendrait une liste vide `[]` :
```

N'aie pas non plus peur de "tableau" pour "array" hein…  C'est un terme tech, la VO dit pas "list" non plus…  Je laisse "liste" ici parce que c'est pas grave, mais tout le reste des trads dit "tableau" partout, pour info.

  const selectedCount = selectedIds.length;

  function handleToggle(toggledId) {

>     // Etait-il sélectionné précédemment ?

Reviewer: ```suggestion
    // Était-il sélectionné avant ?
```

    if (selectedIds.includes(toggledId)) {

>       // Puis retirez l’ID de la liste.

Reviewer: ```suggestion
      // Alors retirer l’ID de la liste.
```

Pas le bon sens de "then"
